web3/contract.py

```diff
@@ -968,14 +953,66 @@ def factory(cls, **kwargs):
         for key in kwargs:
             if not hasattr(cls, key):
                 raise AttributeError(
-                    "Property {0} not found on ContractMethod class. "
-                    "`ContractMethod.factory` only accepts keyword arguments which are "
-                    "present on the ContractMethod class".format(key)
+                    "Property {0} not found on ContractFunction class. "
+                    "`ContractFunction.factory` only accepts keyword arguments which are "
+                    "present on the ContractFunction class".format(key)
                 )
 
         return type(kwargs["method_name"], (cls,), kwargs)
 
 
+class ContractEvent(object):
+    """
+    """
+    address = None
+    event_name = None
+    web3 = None
+    contract_abi = None
+    abi = None
+
+    def __init__(self, *argument_names):
+
+        if argument_names is None:
+            self.argument_names = tuple()
+        else:
+            self.argument_names = argument_names
+
+        self.abi = self._get_event_abi()
+
+    @classmethod
+    def _get_event_abi(cls):
+        return find_matching_event_abi(
+            cls.contract_abi,
+            event_name=cls.event_name)
+
+    def processReceipt(self, txn_receipt):
+        return self._parse_logs(txn_receipt)
+
+    def _parse_logs(self, txn_receipt):
+        parsed_logs = []
+        for log in txn_receipt['logs']:
+            parsed_logs.append(self._decode_log(log))
+        return parsed_logs
+
+    def _decode_log(self, log):
+        return get_event_data(self.abi, log)
+
+    @classmethod
+    def factory(cls, **kwargs):
+        if "event_name" not in kwargs:
+            kwargs["event_name"] = cls.__name__
+
+        for key in kwargs:
+            if not hasattr(cls, key):
+                raise AttributeError(
+                    "Property {0} not found on ContractEvent class. "
```

Comment:
Since this logic is repeated a few times, what do you think about abstracting it away into a `Factory` class in `web3.utils.datatypes`